### DISCUSSION TITLE
Non-MVVM App: Not able to Register with Firebase

### DISCUSSION THREAD

**Discussion by palatinb** [thomasgalliker/Plugin.FirebasePushNotifications#32] [General]
Hello, @thomasgalliker  I would like to ask for some help integrating your package in our app. In our app we do not use MVVM so most of our logic is in the MainPage.xaml.cs. 

**Summarise my problem is:** 
When I would like to call the RegisterForPushNotificationsAsync() method from the MainPage.xaml.cs I get the error that the FirebasePushNotificationManager class logger property is null.

I already added the following line in the MauiProgram.cs file:
```
builder.Services.AddLogging(b =>
	        {
		        b.ClearProviders();
		        b.SetMinimumLevel(LogLevel.Trace);
		        b.AddDebug();
	        });
```
The MainPage.xaml.cs constructor looks like this:
```
public MainPage(IConfiguration config, DeviceRegistrationService deviceRegistrationService,
	        ISafeAreaService safeArea, AppStateService appState, 
	        ILogger<MainPage> logger, IFirebasePushNotification firebasePushNotification,
	        INotificationPermissions notificationPermissions) 

```	        
and it is called from the App.xaml.cs
```
public App(IConfiguration config,
                   DeviceRegistrationService deviceRegistrationService,
                   ISafeAreaService safeAreaService,
                   AppStateService appStateService,
                   IFirebasePushNotification firebasePushNotification,
                   INotificationPermissions notificationPermissions,
                   IServiceProvider serviceProvider)
        {
            InitializeComponent();
            var logger = serviceProvider.GetRequiredService<ILogger<MainPage>>();
            MainPage = new MainPage(config,
                                    deviceRegistrationService,
                                    safeAreaService,
                                    appStateService,
                                    logger,
                                    firebasePushNotification,
                                    notificationPermissions);
        } 
```

Could you please help me resolve this issue? Thanks for your hard work!

**Comment by thomasgalliker** [comment]
Thanks for raising this topic. Before I start to write a long answer: Did you have an look at the [sample app](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/tree/develop/Samples)?

First of all, don't forget to call `.UseFirebasePushNotifications()` in MauiProgram. This call configures the plug-in and registers services (just as `IFirebasePushNotification`). This is very important, since you later want to inject `IFirebasePushNotification `in your MainPage.xaml.cs. 

Do you have a stacktrace of the error that happens when you start the app? Do you have a reproduction sample app which I can use to troubleshoot the problem?

Meanwhile, I'll double-check that there is always a logger instance in FirebasePushNotificationManager...

**Reply by palatinb** [reply]
Yes, I've looked at the Sample App, just we are not using currently the MVVM pattern.

Yes, I call the `UseFirebasePushNotifications()` in the MauiProgram.cs

Currently I do not have a stacktrace, tomorrow I'm going to generate one.

I also tried creating a template app, adding the nuget package then configuring the basic settings, but got the same error.

I'm going to check back tomorrow with the stacktrace!

Thanks for the help!

**Reply by palatinb** [reply]
Hi @thomasgalliker 

Also  I forgot to mention I'm using the latest pre version 2.2.64.
Here is the Stacktrace fort the error:

```
System.ArgumentNullException: Value cannot be null. (Parameter 'logger')
   at System.ThrowHelper.Throw(String paramName)
   at System.ThrowHelper.ThrowIfNull(Object argument, String paramName)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel logLevel, EventId eventId, Exception exception, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel logLevel, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.LogDebug(ILogger logger, String message, Object[] args)
   at Plugin.FirebasePushNotifications.Platforms.FirebasePushNotificationManager.RegisterForPushNotificationsAsync()
   at UniversalPageShell.MainPage.RegisterFirebaseCloud() in /Users/unixdevelop/DEV/WEB/UniversalPageShell-UX/UniversalPageShell/MainPage.xaml.cs:line 144
```

If you need any extra info feel free to contact me

Thanks!

**Reply by palatinb** [reply]
Hello  @thomasgalliker ,

I brought some updates to my problem, We refactored our application to use the MVVM pattern so the app mostly the same as your Sample App. The difference is only that I outsourced the code related to the PushNotification to its own Service. 

When I call the `RegisterForPushNotificationsAsync()` the `IFirebasePushNotification` interface looks like this:
```
 {Plugin.FirebasePushNotifications.Platforms.FirebasePushNotificationManager}
    NotificationCategories: Exception of type 'System.NullReferenceException' was thrown
    NotificationHandler: null
    SubscribedTopics: Exception of type 'System.NullReferenceException' was thrown
    Token: Exception of type 'System.NullReferenceException' was thrown
    disposed (Plugin.FirebasePushNotifications.Platforms.FirebasePushNotificationManagerBase): false
    disposed: false
    hasToken: false
    instanceId: "bb33c"
    logger: null
    notificationActionEventHandler: {System.EventHandler<Plugin.FirebasePushNotifications.FirebasePushNotificationResponseEventArgs>}
    notificationActionQueue: null
    notificationCategories: null
    notificationDeletedEventHandler: {System.EventHandler<Plugin.FirebasePushNotifications.FirebasePushNotificationDataEventArgs>}
    notificationDeletedQueue: null
    notificationOpenedEventHandler: {System.EventHandler<Plugin.FirebasePushNotifications.FirebasePushNotificationResponseEventArgs>}
    notificationOpenedQueue: null
    notificationReceivedEventHandler: {System.EventHandler<Plugin.FirebasePushNotifications.FirebasePushNotificationDataEventArgs>}
    notificationReceivedQueue: null
    pendingTopics: {Count = 0}
    preferences: null
    subscribedTopics: null
    tokenRefreshEventHandler: {System.EventHandler<Plugin.FirebasePushNotifications.FirebasePushNotificationTokenEventArgs>}
    tokenRefreshQueue: null
```

In the Debug Output I also get the following logs:
```
[APP][11:45:50.6735 | DEBUG] FinishedLaunching
Plugin.FirebasePushNotifications.MauiAppBuilderExtensions: Debug: FinishedLaunching
[APP][11:45:50.6756 | DEBUG] Configure
Plugin.FirebasePushNotifications.Platforms.FirebasePushNotificationManager: Debug: Configure
```

So it looks like the `Token` is also not initialised and cannot receive notifications. I hope it helps a bit.

Thanks for the help!

**Reply by thomasgalliker** [reply]
Logger should not be null. Very strange.

**Reply by palatinb** [reply]
Hi @thomasgalliker,

I was able to  resolve our issue. I called it too early and the Service has not been initialised yet. Now it is working correctly, we can receive the notifications, but now we stuck with another problem. 

We added Actions for 1 type of notifications which is working correctly on iOS and open Android also if the app is in the foreground, but if the App is in the background, the notification actions are not showing.

Could you please advise me what can be the problem?

**Reply by thomasgalliker** [reply]
Ok cool. Yes this is probably an error I introduced recently. I changed sonething in the logic how actions are handled and destroyed the whole logic. I have fixed it but did not have the time to test it. I‘ll do it asap (maybe tonight).

**Reply by thomasgalliker** [reply]
Can you try [v2.2.68-pre](https://www.nuget.org/packages/Plugin.FirebasePushNotifications/2.2.68-pre)? This nuget has a refactored handling for Android actions. INotificationBuilder is the new Android-specific interface which builds Android notification UIs for actions. Your feedback is valuable.

**Reply by palatinb** [reply]
Hello @thomasgalliker,

I implemented and tested the NotificationBulder, it is easy to use, but if the application is in the background the notification still missing the actions.

Btw my Firebase message looks like this:
```
{
    Android: {
        CollapseKey: null
        Data: null
        FcmOptions: null
        Notification: {
            Body: null
            BodyLocArgs: null
            BodyLocKey: null
            ChannelId: null
            ClickAction: "2FA_AUTH"
            Color: null
            DefaultLightSettings: false
            DefaultSound: false
            DefaultVibrateTimings: false
            EventTimestamp: {0001. 01. 01. 0:00:00}
            Icon: null
            ImageUrl: null
            LightSettings: null
            LocalOnly: false
            NotificationCount: null
            Priority: null
            Sound: null
            Sticky: false
            Tag: null
            Ticker: null
            Title: null
            TitleLocArgs: null
            TitleLocKey: null
            VibrateTimingsMillis: null
            Visibility: null
        }
    Priority: High
    RestrictedPackageName: null
    TimeToLive: null
    }
    Apns: {
        Aps: {
            Alert: null
            AlertString: null
            Badge: null
            Category: "2FA_AUTH"
            ContentAvailable: false
            CriticalSound: null
            CustomData: null
            MutableContent: false
            Sound: "default"
            ThreadId: null
        }
        CustomData: null
        FcmOptions: null
        Headers: null
    }
    Condition: null
    Data: [
        [0]: {[title, Two Factor Authentication]}
        [1]: {[body, “To Login you need to approve this message”]}
        [2]: {[type, QUESTION_DIALOG]}
        [3]: {[postToWebView, N]}
        [4]: {[imageUrl, ]}
        [5]: {[yesText, Approve]}
        [6]: {[yesUrl, responseUrl]}
        [7]: {[noText, Decline]}
        [8]: {[noUrl, responseUrl]}
        [9]: {[click_action, 2FA_AUTH]}
        [10]: {[actions, [{"title": "Approve", "action": "APPROVE_ACTION"}, {"title": "Decline", "action": "DECLINE_ACTION"}]]}
    ]
    FcmOptions: null
    Notification: {
         Body: “To Login you need to approve this message”
        ImageUrl: null
        Title: "Two Factor Authentication"
    }
    Token: “suppressed”
    Topic: null
    Webpush: null
}
```
